File: api-pytest-framework/agents/planning-agent.py
# ============================================================
# Planning Agent — Jira · Stories · Sprint · Test Cases
# ============================================================
# Absorbe : user-stories-agent · sprint-agent · jira-agent ·
#           jira-ticket-agent · test-case-agent · status-agent · create_story
#
# Commandes :
#   python agents/planning-agent.py setup            → configure le projet Jira (HBAPI)
#   python agents/planning-agent.py stories          → crée les 8 US dans Jira
#   python agents/planning-agent.py sprint list      → liste les sprints
#   python agents/planning-agent.py sprint create "Sprint 2" 2026-07-01 2026-07-21
#   python agents/planning-agent.py sprint start <id>
#   python agents/planning-agent.py sprint close <id>
#   python agents/planning-agent.py tc               → crée les TCs (sous-tâches Jira)
#   python agents/planning-agent.py tickets          → crée des bugs Jira depuis les échecs Allure
#   python agents/planning-agent.py sync             → synchronise Allure → Jira (statuts)
# ============================================================

import sys, os, json, glob, re, time, requests
sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.path.insert(0, os.path.dirname(__file__))

import llm
from jira_fetcher_agent import JiraClient, JIRA_BASE_URL

FRAMEWORK   = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
RESULTS_DIR = os.path.join(FRAMEWORK, "allure-results")
DOCS_DIR    = os.path.join(FRAMEWORK, "docs")

R = "\033[31m"; G = "\033[32m"; Y = "\033[33m"; C = "\033[36m"
W = "\033[1m";  E = "\033[0m"

PROJECT_KEY = "HBAPI"
BOARD_NAME  = "Hotel Booking — API Automation"

# Mapping Allure → Jira transitions
STATUS_MAP = {
    "passed":  "Done",
    "failed":  "In Progress",
    "broken":  "In Progress",
    "skipped": "To Do",
}

# 8 User Stories de la plateforme Hotel Booking API
US_DEFINITIONS = [
    {"id": "US-001", "endpoint": "POST /auth",          "summary": "Token d'authentification",        "priority": "Highest"},
    {"id": "US-002", "endpoint": "GET /booking",        "summary": "Liste des réservations",          "priority": "High"},
    {"id": "US-003", "endpoint": "GET /booking/{id}",   "summary": "Détail d'une réservation",        "priority": "High"},
    {"id": "US-004", "endpoint": "POST /booking",       "summary": "Création d'une réservation",      "priority": "High"},
    {"id": "US-005", "endpoint": "PUT /booking/{id}",   "summary": "Mise à jour complète",            "priority": "Medium"},
    {"id": "US-006", "endpoint": "PATCH /booking/{id}", "summary": "Mise à jour partielle",           "priority": "Medium"},
    {"id": "US-007", "endpoint": "DELETE /booking/{id}","summary": "Suppression d'une réservation",   "priority": "Medium"},
    {"id": "US-008", "endpoint": "GET /ping",           "summary": "Health check de l'API",           "priority": "Low"},
]


# ── Helpers ────────────────────────────────────────────────────────────────

def jira_post(path: str, payload: dict, jira: JiraClient) -> dict:
    resp = requests.post(
        f"{JIRA_BASE_URL}{path}",
        json=payload, auth=jira.auth, headers=jira.headers, verify=False
    )
    return resp.json() if resp.status_code in (200, 201) else {"error": resp.text[:200], "status": resp.status_code}


def jira_get(path: str, jira: JiraClient) -> dict:
    resp = requests.get(
        f"{JIRA_BASE_URL}{path}",
        auth=jira.auth, headers=jira.headers, verify=False
    )
    return resp.json() if resp.status_code == 200 else {}


def load_allure_results() -> list:
    results = []
    for f in glob.glob(os.path.join(RESULTS_DIR, "*-result.json")):
        try:
            d = json.load(open(f, encoding="utf-8"))
            tags  = [lb["value"] for lb in d.get("labels", []) if lb["name"] == "tag"]
            tc    = next((t for t in tags if re.match(r"tc-\d+", t)), None)
            us    = next((t for t in tags if re.match(r"us-\d+", t)), None)
            results.append({
                "name":   d.get("name", "?"),
                "status": d.get("status", "unknown"),
                "tc":     tc, "us": us,
                "message": (d.get("statusDetails") or {}).get("message", "")[:200],
            })
        except Exception:
            pass
    return results


# ── Setup — Configuration Jira ─────────────────────────────────────────────

def cmd_setup():
    print(f"\n{W}PLANNING AGENT — Setup Jira{E}\n")
    jira = JiraClient()

    # Vérifier la connexion
    myself = jira_get("/rest/api/3/myself", jira)
    if "accountId" not in myself:
        print(f"{R}  [ERR] Jira inaccessible. Verifiez JIRA_URL / JIRA_EMAIL / JIRA_TOKEN dans .env{E}")
        return

    print(f"  {G}Connecte en tant que : {myself.get('displayName','?')}{E}")

    # Créer le projet si absent
    projects = jira_get("/rest/api/3/project/search?query=HBAPI", jira)
    existing = [p for p in projects.get("values", []) if p.get("key") == PROJECT_KEY]

    if existing:
        print(f"  {Y}Projet {PROJECT_KEY} existe deja.{E}")
    else:
        print(f"  Creation du projet {PROJECT_KEY}...")
        payload = {
            "key":          PROJECT_KEY,
            "name":         BOARD_NAME,
            "projectTypeKey": "software",
            "template":     "com.pyxis.greenhopper.jira:gh-scrum-template",
        }
        result = jira_post("/rest/api/3/project", payload, jira)
        if "key" in result:
            print(f"  {G}Projet cree : {result['key']}{E}")
        else:
            print(f"  {Y}Erreur creation : {result}{E}")


# ── Stories — Création des User Stories ────────────────────────────────────

def create_user_story(us: dict, jira: JiraClient) -> str:
    # Générer la description via LLM
    messages = [{"role": "user", "content": (
        f"Génère une user story Jira pour cet endpoint API :\n\n"
        f"ID       : {us['id']}\n"
        f"Endpoint : {us['endpoint']}\n"
        f"Résumé   : {us['summary']}\n\n"
        f"Format : 'En tant que... je veux... afin de...' + critères d'acceptation (3 points).\n"
        f"Max 300 caractères. En français."
    )}]
    description_text = llm.chat(messages)

    payload = {
        "fields": {
            "project":     {"key": PROJECT_KEY},
            "summary":     f"[{us['id']}] {us['endpoint']} — {us['summary']}",
            "description": {
                "type": "doc", "version": 1,
                "content": [{"type": "paragraph", "content": [{"type": "text", "text": description_text}]}]
            },
            "issuetype": {"name": "Story"},
            "priority":  {"name": us["priority"]},
            "labels":    [us["id"].lower().replace("-", ""), "api-automation"],
        }
    }
    result = jira_post("/rest/api/3/issue", payload, jira)
    return result.get("key", "")


def cmd_stories(dry_run: bool = False):
    print(f"\n{W}PLANNING AGENT — User Stories ({len(US_DEFINITIONS)} US){E}\n")
    jira = JiraClient()
    created = []

    for us in US_DEFINITIONS:
        print(f"  {C}[{us['id']}]{E} {us['endpoint']:<25}  {us['summary']}...", end=" ", flush=True)
        if dry_run:
            print(f"{Y}[DRY-RUN]{E}")
            continue
        key = create_user_story(us, jira)
        if key:
            created.append(key)
            print(f"{G}{key}{E}")
        else:
            print(f"{R}Erreur{E}")
        time.sleep(0.5)  # Rate limiting

    print(f"\n  {G}{len(created)} US creees.{E}")
    return created


# ── Sprint — Gestion des sprints ───────────────────────────────────────────

def get_board_id(jira: JiraClient) -> str:
    boards = jira_get(f"/rest/agile/1.0/board?projectKeyOrId={PROJECT_KEY}", jira)
    values = boards.get("values", [])
    return str(values[0]["id"]) if values else ""


def cmd_sprint(sub: str = "list", name: str = None, start: str = None, end: str = None, goal: str = None, sprint_id: str = None):
    print(f"\n{W}PLANNING AGENT — Sprint [{sub}]{E}\n")
    jira = JiraClient()

    if sub == "list":
        board_id = get_board_id(jira)
        if not board_id:
            print(f"{R}  Board non trouve pour {PROJECT_KEY}.{E}")
            return
        sprints = jira_get(f"/rest/agile/1.0/board/{board_id}/sprint?state=active,future,closed&maxResults=10", jira)
        for s in sprints.get("values", []):
            state = s.get("state", "?")
            color = G if state == "active" else Y if state == "future" else C
            print(f"  #{s['id']}  {color}{state:<8}{E}  {s.get('name','?'):<30}  {s.get('startDate','?')[:10]} → {s.get('endDate','?')[:10]}")

    elif sub == "create" and name:
        board_id = get_board_id(jira)
        payload = {
            "name":      name,
            "startDate": f"{start}T00:00:00.000Z" if start else "",
            "endDate":   f"{end}T23:59:59.000Z" if end else "",
            "goal":      goal or f"Objectif du {name}",
            "originBoardId": int(board_id),
        }
        result = jira_post("/rest/agile/1.0/sprint", payload, jira)
        if "id" in result:
            print(f"  {G}Sprint cree : #{result['id']} — {name}{E}")
        else:
            print(f"  {R}Erreur : {result}{E}")

    elif sub in ("start", "close") and sprint_id:
        state_map = {"start": "active", "close": "closed"}
        payload   = {"state": state_map[sub]}
        resp = requests.post(
            f"{JIRA_BASE_URL}/rest/agile/1.0/sprint/{sprint_id}",
            json=payload, auth=jira.auth, headers=jira.headers, verify=False
        )
        if resp.status_code in (200, 204):
            print(f"  {G}Sprint #{sprint_id} -> {state_map[sub]}{E}")
        else:
            print(f"  {R}Erreur {resp.status_code}: {resp.text[:100]}{E}")


# ── TC — Test Cases Jira ───────────────────────────────────────────────────

def cmd_tc(us_filter: str = None, dry_run: bool = False):
    print(f"\n{W}PLANNING AGENT — Test Cases{E}\n")
    jira   = JiraClient()
    us_list = [us for us in US_DEFINITIONS if not us_filter or us["id"] == us_filter.upper()]

    for us in us_list:
        us_key = us["id"].lower().replace("-", "")
        print(f"\n  {C}[{us['id']}]{E} {us['endpoint']} — {us['summary']}")

        # Générer les TCs via LLM
        messages = [{"role": "user", "content": (
            f"Génère 3 cas de test pour cet endpoint API :\n\n"
            f"Endpoint : {us['endpoint']}\n"
            f"Résumé   : {us['summary']}\n\n"
            f"Format JSON : {{\"test_cases\": [{{\"title\": \"...\", \"type\": \"positif|negatif|securite\", \"steps\": [\"...\"], \"expected\": \"...\"}}]}}"
        )}]
        try:
            schema = {
                "type": "object",
                "properties": {"test_cases": {"type": "array", "items": {"type": "object",
                    "properties": {"title": {"type": "string"}, "type": {"type": "string"},
                                   "steps": {"type": "array", "items": {"type": "string"}}, "expected": {"type": "string"}},
                    "required": ["title", "type", "expected"]}}},
                "required": ["test_cases"]
            }
            result = llm.chat_structured(messages, schema)
            test_cases = result.get("test_cases", [])
        except Exception:
            test_cases = []

        for tc in test_cases:
            title = f"[{us['id']}][{tc.get('type','?').upper()}] {tc.get('title','?')}"
            print(f"    {Y}TC{E}  {title[:70]}", end=" ", flush=True)
            if dry_run:
                print(f"{Y}[DRY-RUN]{E}")
                continue

            # Chercher l'issue US parente
            jql_result = jira_get(
                f"/rest/api/3/issue/picker?query=[{us['id']}]&currentJQL=project={PROJECT_KEY}", jira
            )
            parent_key = ""

            payload = {
                "fields": {
                    "project":     {"key": PROJECT_KEY},
                    "summary":     title,
                    "issuetype":   {"name": "Subtask" if parent_key else "Task"},
                    "labels":      [us_key, "tc", tc.get("type", "")],
                }
            }
            if parent_key:
                payload["fields"]["parent"] = {"key": parent_key}

            tc_result = jira_post("/rest/api/3/issue", payload, jira)
            if "key" in tc_result:
                print(f"{G}{tc_result['key']}{E}")
            else:
                print(f"{R}Erreur{E}")
            time.sleep(0.3)


# ── Tickets — Bugs depuis les échecs Allure ────────────────────────────────

def cmd_tickets(dry_run: bool = False):
    print(f"\n{W}PLANNING AGENT — Bug Tickets depuis Allure{E}\n")
    jira    = JiraClient()
    results = load_allure_results()
    failures = [r for r in results if r["status"] in ("failed", "broken")]

    if not failures:
        print(f"{G}  Aucun echec Allure — aucun ticket a creer.{E}")
        return

    print(f"  {len(failures)} echec(s) -> tickets Jira\n")
    created = []

    for f in failures:
        # Générer le titre du bug via LLM
        messages = [{"role": "user", "content": (
            f"Génère un titre de bug Jira clair et concis (max 80 chars, en français).\n\n"
            f"Test en echec : {f['name']}\n"
            f"TC           : {f['tc'] or 'N/A'}\n"
            f"Erreur       : {f['message'] or 'aucun message'}\n\n"
            f"Format : [TC-XXX] Description courte du bug"
        )}]
        summary = llm.chat(messages).strip()[:120]

        print(f"  {R}x{E} {f['tc'] or '?':>8}  {summary[:60]}", end=" ", flush=True)
        if dry_run:
            print(f"{Y}[DRY-RUN]{E}")
            continue

        payload = {
            "fields": {
                "project":     {"key": PROJECT_KEY},
                "summary":     summary,
                "description": {
                    "type": "doc", "version": 1,
                    "content": [{"type": "paragraph", "content": [{"type": "text", "text": (
                        f"Test: {f['name']}\n"
                        f"TC: {f['tc'] or 'N/A'} | US: {f['us'] or 'N/A'}\n"
                        f"Statut: {f['status']}\n\n"
                        f"Erreur:\n{f['message'] or 'Aucun message'}"
                    )}]}]
                },
                "issuetype": {"name": "Bug"},
                "priority":  {"name": "High"},
                "labels":    ["automated-failure", f['tc'] or "no-tc"],
            }
        }
        result = jira_post("/rest/api/3/issue", payload, jira)
        if "key" in result:
            created.append(result["key"])
            print(f"{G}{result['key']}{E}")
        else:
            print(f"{R}Erreur{E}")
        time.sleep(0.5)

    print(f"\n  {G}{len(created)} ticket(s) Bug crees.{E}")
    return created


# ── Sync — Allure → Jira statuts ──────────────────────────────────────────

def cmd_sync():
    print(f"\n{W}PLANNING AGENT — Sync Allure → Jira{E}\n")
    jira    = JiraClient()
    results = load_allure_results()

    synced = 0
    for r in results:
        tc  = r.get("tc")
        if not tc:
            continue

        allure_status = r.get("status", "unknown")
        jira_status   = STATUS_MAP.get(allure_status)
        if not jira_status:
            continue

        # Chercher le TC dans Jira par label
        search = jira_get(f"/rest/api/3/issue/picker?query={tc}&currentJQL=project={PROJECT_KEY}+AND+labels={tc}", jira)
        issues = search.get("sections", [{}])[0].get("issues", []) if search else []

        for issue in issues[:1]:
            issue_key = issue.get("key")
            if not issue_key:
                continue
            transitions = jira_get(f"/rest/api/3/issue/{issue_key}/transitions", jira)
            target = next(
                (t for t in transitions.get("transitions", []) if jira_status.lower() in t.get("name","").lower()),
                None
            )
            if target:
                resp = requests.post(
                    f"{JIRA_BASE_URL}/rest/api/3/issue/{issue_key}/transitions",
                    json={"transition": {"id": target["id"]}},
                    auth=jira.auth, headers=jira.headers, verify=False
                )
                if resp.status_code in (200, 204):
                    print(f"  {G}{tc}{E} {issue_key} -> {jira_status}")
                    synced += 1

    print(f"\n  {G}{synced} issue(s) synchronisee(s).{E}")


# ── Main ───────────────────────────────────────────────────────────────────

def print_help():
    print(f"""
{W}PLANNING AGENT — Jira · User Stories · Sprint · Test Cases{E}

  python agents/planning-agent.py setup                  Configure le projet Jira (HBAPI)
  python agents/planning-agent.py stories                Crée les 8 US dans Jira
  python agents/planning-agent.py stories --dry-run      Simulation
  python agents/planning-agent.py sprint list            Liste les sprints
  python agents/planning-agent.py sprint create "S2" 2026-07-01 2026-07-21
  python agents/planning-agent.py sprint start <id>     Démarre un sprint
  python agents/planning-agent.py sprint close <id>     Clôture un sprint
  python agents/planning-agent.py tc                     Crée tous les TCs dans Jira
  python agents/planning-agent.py tc US-001             TCs d'une seule US
  python agents/planning-agent.py tickets                Crée les bugs Jira depuis Allure
  python agents/planning-agent.py tickets --dry-run      Simulation
  python agents/planning-agent.py sync                   Synchronise statuts Allure → Jira

{W}Modules absorbes :{E} user-stories-agent · sprint-agent · jira-agent · jira-ticket-agent · test-case-agent · status-agent · create_story
""")


if __name__ == "__main__":
    dry = "--dry-run" in sys.argv
    cmd = sys.argv[1] if len(sys.argv) > 1 else "help"
    sub = sys.argv[2] if len(sys.argv) > 2 and not sys.argv[2].startswith("-") else None
    arg = sys.argv[3] if len(sys.argv) > 3 and not sys.argv[3].startswith("-") else None
    arg2 = sys.argv[4] if len(sys.argv) > 4 else None
    arg3 = sys.argv[5] if len(sys.argv) > 5 else None

    if cmd == "setup":
        cmd_setup()
    elif cmd == "stories":
        cmd_stories(dry_run=dry)
    elif cmd == "sprint":
        cmd_sprint(sub or "list", name=sub if sub != "list" else arg,
                   start=arg, end=arg2, goal=arg3, sprint_id=arg)
    elif cmd == "tc":
        cmd_tc(us_filter=sub, dry_run=dry)
    elif cmd == "tickets":
        cmd_tickets(dry_run=dry)
    elif cmd == "sync":
        cmd_sync()
    else:
        print_help()

File: ui_cypress_bdd/scripts/agents/planning-agent.js
'use strict';
// ============================================================
// Planning Agent — Gestion projet & couverture
// ============================================================
// Usage:
//   node scripts/agents/planning-agent.js stories              Liste les US Jira
//   node scripts/agents/planning-agent.js stories create       Crée les US dans Jira
//   node scripts/agents/planning-agent.js testcases            Liste les TCs existants
//   node scripts/agents/planning-agent.js sprint list          Liste les sprints
//   node scripts/agents/planning-agent.js sprint create <nom>  Crée un sprint
//   node scripts/agents/planning-agent.js sprint start <id>    Démarre un sprint
//   node scripts/agents/planning-agent.js sprint close <id>    Ferme un sprint
//   node scripts/agents/planning-agent.js coverage analyze     Analyse la couverture .feature
//   node scripts/agents/planning-agent.js coverage gaps        Identifie les manques
//   node scripts/agents/planning-agent.js coverage suggest     Suggestions LLM
//
// Coût LLM : bas à moyen — seulement sur coverage suggest
// ============================================================
require('dotenv').config();
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const fs   = require('fs');
const path = require('path');
const https = require('https');
const llm  = require('./llm');
const jira = require('./jira-fetcher');
const tracer = require('./shared/tracer');

const FRAMEWORK    = path.join(__dirname, '..', '..');
const FEATURES_DIR = path.join(FRAMEWORK, 'cypress', 'e2e', 'features');
const DOCS_DIR      = path.join(FRAMEWORK, 'docs');

fs.mkdirSync(DOCS_DIR, { recursive: true });

const DRY_RUN = process.argv.includes('--dry-run');
const PROJECT = process.env.JIRA_PROJECT || 'SCRUM';
const JIRA_BASE_URL = (process.env.JIRA_BASE_URL || '').replace(/\/$/, '');
const JIRA_AUTH = Buffer.from(`${process.env.JIRA_EMAIL}:${process.env.JIRA_TOKEN}`).toString('base64');

const G = '\x1b[32m', R = '\x1b[31m', Y = '\x1b[33m', C = '\x1b[36m', B = '\x1b[1m', E = '\x1b[0m';

// ── Agile API helper ───────────────────────────────────────────────────────────
function agileGet(endpoint) {
  return new Promise((resolve, reject) => {
    const url = `${JIRA_BASE_URL}/rest/agile/1.0${endpoint}`;
    https.get(url, { headers: { Authorization: `Basic ${JIRA_AUTH}`, Accept: 'application/json' } }, res => {
      let d = ''; res.on('data', c => d += c);
      res.on('end', () => { try { resolve({ status: res.statusCode, body: JSON.parse(d) }); } catch { reject(new Error(d.slice(0,200))); } });
    }).on('error', reject);
  });
}

function agilePost(endpoint, body) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(body);
    const url  = new URL(`${JIRA_BASE_URL}/rest/agile/1.0${endpoint}`);
    const req  = https.request({ hostname: url.hostname, path: url.pathname, method: 'POST',
      headers: { Authorization: `Basic ${JIRA_AUTH}`, 'Content-Type': 'application/json', Accept: 'application/json', 'Content-Length': Buffer.byteLength(data) }
    }, res => {
      let d = ''; res.on('data', c => d += c);
      res.on('end', () => { try { resolve({ status: res.statusCode, body: d ? JSON.parse(d) : {} }); } catch { resolve({ status: res.statusCode, body: d }); } });
    });
    req.on('error', reject); req.write(data); req.end();
  });
}

// ── STORIES ───────────────────────────────────────────────────────────────────
async function cmdStories(action) {
  console.log(`\n${B}=== PLANNING — STORIES ===${E}`);
  try {
    await jira.assertJira();
    const stories = await jira.fetchStories(PROJECT, 100);

    if (action === 'create') {
      console.log(`  ${Y}⚠  Création manuelle requise — utilisez Jira UI ou jira-ticket-agent${E}`);
      console.log(`  URL Backlog : ${JIRA_BASE_URL}/jira/software/projects/${PROJECT}/boards`);
      return;
    }

    console.log(`  ${C}${stories.length} User Stories dans ${PROJECT}${E}\n`);
    console.log(`  ${'Clé'.padEnd(12)} ${'Statut'.padEnd(15)} ${'Priorité'.padEnd(10)} Titre`);
    console.log(`  ${'─'.repeat(80)}`);
    for (const s of stories) {
      const statusColor = s.status === 'Done' ? G : s.status === 'In Progress' ? C : '';
      console.log(`  ${s.key.padEnd(12)} ${statusColor}${s.status.padEnd(15)}${E} ${s.priority.padEnd(10)} ${s.summary.slice(0,45)}`);
    }
  } catch (e) {
    console.error(`  ${R}✗ Jira: ${e.message}${E}`);
  }
}

// ── SPRINT ────────────────────────────────────────────────────────────────────
async function cmdSprint(action, arg) {
  console.log(`\n${B}=== PLANNING — SPRINT ${action?.toUpperCase()||''} ===${E}`);
  try {
    const { body: boardBody } = await agileGet(`/board?projectKeyOrId=${PROJECT}`);
    const board = boardBody.values?.[0];
    if (!board) { console.log(`  ${Y}⚠  Aucun board Agile trouvé pour ${PROJECT}${E}`); return; }
    const boardId = board.id;

    if (action === 'list' || !action) {
      const { body } = await agileGet(`/board/${boardId}/sprint?state=active,future,closed&maxResults=20`);
      const sprints  = body.values || [];
      console.log(`  ${C}Board : ${board.name}${E}\n`);
      sprints.forEach(s => {
        const stateColor = s.state === 'active' ? G : s.state === 'future' ? C : '\x1b[2m';
        console.log(`  ${stateColor}[${s.state.toUpperCase()}]${E}  #${s.id}  ${s.name}  ${s.startDate?.slice(0,10)||''}→${s.endDate?.slice(0,10)||''}`);
      });
    } else if (action === 'create') {
      const name = arg || `Sprint ${new Date().toLocaleDateString('fr-FR')}`;
      if (DRY_RUN) { console.log(`  ${Y}[DRY-RUN] Aurait créé: ${name}${E}`); return; }
      const { status, body } = await agilePost(`/sprint`, { name, originBoardId: boardId });
      if (status === 201) console.log(`  ${G}✓ Sprint créé : #${body.id} — ${name}${E}`);
      else console.error(`  ${R}✗ HTTP ${status}: ${JSON.stringify(body).slice(0,100)}${E}`);
    } else if (action === 'start') {
      if (!arg) { console.log(`  Usage: planning sprint start <id>`); return; }
      if (DRY_RUN) { console.log(`  ${Y}[DRY-RUN] Aurait démarré sprint #${arg}${E}`); return; }
      const start = new Date(); const end = new Date(start); end.setDate(end.getDate()+14);
      const { status } = await new Promise((resolve, reject) => {
        const data = JSON.stringify({ state: 'active', startDate: start.toISOString(), endDate: end.toISOString() });
        const url  = new URL(`${JIRA_BASE_URL}/rest/agile/1.0/sprint/${arg}`);
        const req  = https.request({ hostname: url.hostname, path: url.pathname, method: 'POST',
          headers: { Authorization: `Basic ${JIRA_AUTH}`, 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) }
        }, res => resolve({ status: res.statusCode }));
        req.on('error', reject); req.write(data); req.end();
      });
      console.log(status < 300 ? `  ${G}✓ Sprint #${arg} démarré${E}` : `  ${R}✗ HTTP ${status}${E}`);
    } else if (action === 'close') {
      if (!arg) { console.log(`  Usage: planning sprint close <id>`); return; }
      if (DRY_RUN) { console.log(`  ${Y}[DRY-RUN] Aurait fermé sprint #${arg}${E}`); return; }
      const data = JSON.stringify({ state: 'closed' });
      const url  = new URL(`${JIRA_BASE_URL}/rest/agile/1.0/sprint/${arg}`);
      const { status } = await new Promise((resolve, reject) => {
        const req = https.request({ hostname: url.hostname, path: url.pathname, method: 'POST',
          headers: { Authorization: `Basic ${JIRA_AUTH}`, 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) }
        }, res => resolve({ status: res.statusCode }));
        req.on('error', reject); req.write(data); req.end();
      });
      console.log(status < 300 ? `  ${G}✓ Sprint #${arg} fermé${E}` : `  ${R}✗ HTTP ${status}${E}`);
    }
  } catch (e) {
    console.error(`  ${R}✗ ${e.message}${E}`);
  }
}

// ── COVERAGE ──────────────────────────────────────────────────────────────────
const COVERAGE_TYPES = ['positif', 'negatif', 'auth', 'limite', 'securite', 'performance'];

async function cmdCoverage(action) {
  console.log(`\n${B}=== PLANNING — COVERAGE ===${E}`);
  if (!fs.existsSync(FEATURES_DIR)) {
    console.log(`  ${Y}⚠  Aucun .feature trouvé dans cypress/e2e/features/${E}`); return;
  }

  const features = fs.readdirSync(FEATURES_DIR).filter(f => f.endsWith('.feature'));
  const coverage = {};

  for (const f of features) {
    const content = fs.readFileSync(path.join(FEATURES_DIR, f), 'utf8');
    const scenarios = (content.match(/Scenario:/g) || []).length;
    const tags      = (content.match(/@[\w-]+/g) || []).map(t => t.slice(1));
    const types     = COVERAGE_TYPES.filter(t => tags.includes(t));
    coverage[f] = { scenarios, types, tags };
  }

  console.log(`  ${C}${features.length} features analysées${E}\n`);

  if (action === 'analyze' || !action) {
    let total = 0;
    features.forEach(f => {
      const c = coverage[f];
      total += c.scenarios;
      const typeStr = c.types.length ? c.types.join(', ') : Y+'aucun type détecté'+E;
      console.log(`  ${G}✓${E} ${f.padEnd(35)} ${String(c.scenarios).padEnd(4)} scénarios  [${typeStr}]`);
    });
    console.log(`\n  Total : ${total} scénarios dans ${features.length} features`);
  } else if (action === 'gaps') {
    console.log(`  ${B}Couverture manquante par feature :${E}\n`);
    features.forEach(f => {
      const missing = COVERAGE_TYPES.filter(t => !coverage[f].types.includes(t));
      if (missing.length) console.log(`  ${Y}⚠${E}  ${f.padEnd(35)} manque: ${R}${missing.join(', ')}${E}`);
      else console.log(`  ${G}✓${E}  ${f.padEnd(35)} couverture complète`);
    });
  } else if (action === 'suggest') {
    const gaps = features.flatMap(f => {
      const missing = COVERAGE_TYPES.filter(t => !coverage[f].types.includes(t));
      return missing.map(t => ({ feature: f, missing: t }));
    });

    if (!gaps.length) { console.log(`  ${G}✓ Couverture complète — aucune suggestion nécessaire${E}`); return; }
    console.log(`  ${C}Génération suggestions LLM pour ${gaps.length} manques...${E}\n`);

    const span = new tracer.Span('coverageSuggest', JSON.stringify(gaps), llm.MODEL).begin();
    try {
      const prompt = `Tu es un expert QA BDD. Suggère des scénarios Gherkin manquants.

Manques identifiés:
${gaps.map(g => `- ${g.feature}: manque couverture ${g.missing}`).join('\n')}

Pour chaque manque, génère 1 Scenario Gherkin concis en français avec tags appropriés.
Format: @feature @type\nScenario: ...\n  Given...\n  When...\n  Then...`;

      const resp = await llm.chat([{ role: 'user', content: prompt }]);
      span.response = resp.message.content; span.end(true);
      console.log(resp.message.content);
    } catch (e) {
      span.error = e.message; span.end(false);
      console.error(`  ${R}✗ ${e.message}${E}`);
    }
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  const args = process.argv.slice(2).filter(a => !a.startsWith('--'));
  const [cmd, sub, arg3] = args;
  await llm.assertRunning();

  switch (cmd) {
    case 'stories':   await cmdStories(sub);         break;
    case 'sprint':    await cmdSprint(sub, arg3);     break;
    case 'coverage':  await cmdCoverage(sub);         break;
    default:
      console.log(`
${B}Planning Agent${E} — Gestion projet & couverture

  stories [create]         Liste ou crée les User Stories Jira
  sprint list              Liste les sprints du board
  sprint create [nom]      Crée un nouveau sprint
  sprint start <id>        Démarre un sprint
  sprint close <id>        Ferme un sprint
  coverage analyze         Analyse la couverture des .feature
  coverage gaps            Identifie les types de test manquants
  coverage suggest         Suggestions LLM pour combler les manques

Options:
  --dry-run    Simulation sans écriture

Variables:
  JIRA_PROJECT=SCRUM       Projet Jira cible
`);
  }
}

main().catch(e => { console.error(R + e.message + E); process.exit(1); });
